Issue #141 - team 5

premik91 opened the pull request:
A simple design for list of online users. I also added the toggle animation for hidding it and the title for this 'box' (Active Users or Logged in Users doesn't really sound good, but that's the least of my worries :)).

Also added the ability to display only users that contain the searched string and to arrange list of logged in users by alphabet(design needs to be improved). Also the true image of the user should be passed in 'data' to javascript. We (I and igypigi->did most of the javascript) were also unable to test this completely, because the updating logged in users doesn't always work(on our computers) and we weren't able to figure out why.

Commit:
Issue #141 - team 5

A simple design for list of online users. I also added the toggle
animation for hidding it and the title for this 'box' (Active Users or
Logged in Users doesn't really sound good, but that's the least of my
worries :)).

Also added the ability to display only users that contain the searched
string and to arrange list of logged in users by alphabet(design needs
to be improved). Also the true image of the user should be passed in
'data' to javascript. We (I and igypigi->did most of the javascript)
were also unable to test this completely, because the updating logged
in users doesn't always work(on our computers) and we weren't able to
figure out why.

Commit:
Fixing tasks.py

Commit:
Fixing imports in tasks.py

premik91 commented:
Old thread: https://github.com/wlanslovenija/PiplMesh/pull/157

Commit:
Added translation tags

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,67 @@
```
For JavaScript, I recommend you to run some JavaScript lint tool over it. Will tell you even more JavaScript recommendations than me. :-)

mitar commented on piplmesh/frontend/tasks.py:
```diff
@@ -1,11 +1,7 @@
 import datetime
-
```
Why? removing those empty lines? We have them to separate import sections (system modules, django modules, library modules, our modules).

premik91 replied:
Sorry I didn't know we were doing it that way, thought it was a mistake.

mitar replied:
Then you could check in other files. :-)

premik91 replied:
That's the funny thing. I did but I probably opened those that aren't seperated :)

mitar replied:
We have those? Where?

premik91 replied:
No I am not saying they are wrong. I just randomly opened some .py files to see if all imports have one empty line between them. And those that I opened were probably all from the same section. I remeber opening signals.py and some others.

igypigi replied:
{% for user in online_users %}
            <li>{{ user }}</li>
        {% endfor %}
why do we need this? how do I update userlist?

mitar replied:
This was for initial display. You should move this to JavaScript now.

mitar commented on piplmesh/frontend/tasks.py:
```diff
@@ -23,11 +19,14 @@ def check_online_users():
             updates.send_update(
                 views.HOME_CHANNEL_ID,
-                {
+                    {
```
Why changing indentation?

premik91 replied:
I didn't, probably Pycharm did, I will fix it.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,67 @@
+    if (data.action == 'JOIN') {
+        users[users.length] = data;
+        if (alphabeticalOrder) {
+            rebuildUserList();
```
You didn't yet change JavaScript in the way I commented. Will wait for that.

premik91 replied:
Yeah, I told igypigi to fix it(since he wrote most of it, he should fix it).

Commit:
Unrolling "fixes"

Commit:
form action set to ""

Commit:
Fixed indentation in tasks.py

premik91 commented:
I think the rest is up to igypigi :)

mitar commented:
Please update from main repository. Now list should work as expected.

mitar commented:
So you should be able to test everything.

Commit:
fixed confilcts

Commit:
Added translation load tag

mitar commented:
So, does it work now for you?

premik91 commented:
Yeah it does. Excelent! :) 
I will wait for igypigi one more day, than I am gonna do it.

mitar commented on piplmesh/frontend/static/piplmesh/css/style.css:
```diff
@@ -346,4 +346,54 @@ input.gsc-search-button {
+    color: #006600;
+}
+
+/* BOXES */
```
Style this comment like others. Renamed boxes to panels.

premik91 replied:
Fixed

mitar commented on piplmesh/frontend/static/piplmesh/css/style.css:
```diff
@@ -346,4 +346,54 @@ input.gsc-search-button {
+    width: 260px;
+}
+
+.boxHeader {
```
`.boxHeader` -> `.panel .header`?

premik91 replied:
Fixed

mitar commented on piplmesh/frontend/templates/userlist.html:
```diff
@@ -1,7 +1,17 @@
+    <form action="">
+        <img src="{{ STATIC_URL }}piplmesh/images/magnifying_glass.png" alt="magnifying_glass" />
+        <input type="text" id="search_users" />
+        <label>
```
Style this label and input same as we have style elsewhere. (for attribute for example).

premik91 replied:
Fixed

mitar commented:
It would be great if it would be done until tomorrow so that others can see it. Can you call him?

premik91 commented:
He had to take his dog out, and than he is going to sleep.
He says he is gonna do it between the lecture. :)

mitar commented:
I will be beaten by the professor if you will be doing things during the lecture.

premik91 commented:
We will be discreet :)

Commit:
Fixed css

Commit:
Added label attributre

Commit:
Fix from igypigi

mitar commented on piplmesh/frontend/templates/userlist.html:
```diff
@@ -1,7 +1,17 @@
+        <input type="text" id="search_users" />
+        <label for="alphabet_order">
+            <input type="checkbox" id="alphabet_order" />A-Z
+        </label>
```
We have like:

```
<input type="checkbox" id="alphabet_order" /><label for="alphabet_order">A-Z</label>
```

mitar commented on piplmesh/frontend/templates/userlist.html:
```diff
@@ -1,7 +1,17 @@
+
+<div class="panel">
+    <div class="header" title="{% trans "Show/Hide Users" %}">{% trans "Active Users" %}</div>
     <ul id="userlist">
```
You should move this to JavaScript, some global variable. And initialize from there.

mitar commented on piplmesh/frontend/tasks.py:
```diff
@@ -27,6 +27,9 @@ def check_online_users():
                     'username': user.username,
+                    # TODO get user image
+                    'image': '',
+                    'id': user.id,
```
Why we need user ID exposed?

premik91 replied:
We don't :)

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,61 @@
+// TODO add users already online to this array
+var users = new Array();
+var searchUsers;
+var alphabeticalOrder = false;
```
Why would we have non-alphabetical order? I would remove this, and always sort.

igypigi replied:
premik91 said users can be ordered by both time of login or alphabet.

mitar replied:
Simplify. I would use only by alphabet.

premik91 replied:
So we don't need the checkbox?

mitar replied:
We do not.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,61 @@
+        users[users.length] = data;
+        redrawUserList();
+    } else if (data.action == 'PART') {
+        var index = users.indexOf(data);
```
Use jQuery.grep to remove the user from the array.

mitar replied:
Or, if we will have dictionary, you can easier remove it.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,61 @@
+var userList;
+function updateUserList(data) {
+    if (data.action == 'JOIN') {
+        users[users.length] = data;
```
I like `users.push(data)` more.

BTW, you have to check if user is already in there.

mitar replied:
Create User class:

```
function User(data) {
    var self = this;

    $.extend(self, data);

    self.someMethod = function () {
        ...
    };
}
```

And then you do:

```
var user = new User(data.user);
onlineUsers[user.username] = user;
```

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,61 @@
-		$('#userlist li:contains(' + escape(data.username) + ')').remove();
-	}
+// TODO add users already online to this array
+var users = new Array();
```
I would have a dictionary here:

```
var onlineUsers = {};
```

mitar commented on piplmesh/frontend/tasks.py:
```diff
@@ -27,6 +27,9 @@ def check_online_users():
                     'type': 'userlist',
                     'action': 'JOIN',
```
Change this structure to:

```
{
    'type': 'userlist',
    'action': 'JOIN',
    'user': {
        'username': user.username,
        'profile_url': urlresolvers.reverse('user', kwargs={'username': user.username}),
        'image_url': '',
    },
}
```

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,61 @@
+
+// redraws userList
+function redrawUserList() {
+    var tmpUsers = users.slice();
```
```
var users = [];
for (var user in onlineUsers) {
    if (onlineUsers.hasOwnProperty(user)) {
        users.push(user);
    }
}
users.sort(function (user1, user) {
    if (user1.username.toLowerCase() < user2.username.toLowerCase()) return -1;
    if (user1.username.toUpperCase() > user2.username.toUpperCase()) return 1;
    return 0;
}
```

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,61 @@
+    }
+    for (user in tmpUsers) {
+        if (user.username.indexOf(searchUsers) != -1) {
+            userlist.append("<li>" +
```
`userList` != `userlist`

mitar replied:
Don't concatenate strings, but use jQuery to build it (use `'`, not `"`):

```
var image = $('<img/>').attrs({'src': user.image_url, 'alt': _("User image")});
$('<li/>').append(image);
```

and so on ...

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,61 @@
 $(document).ready(function () {
-	$.updates.registerProcessor('home_channel', 'userlist', updateUserlist);
-});
+    userList = $("#userlist");
```
I am not sure if we need to store this. Because query by ID is very fast.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,61 @@
+// redraws userList
+function redrawUserList() {
+    var tmpUsers = users.slice();
+    userList.empty();
```
Move this lower.

mitar commented:
OK. Enough for now. Include this your template somewhere in `home.html`. Pull from main repository new changes.

Commit:
FIrst fixes

Commit:
Fixed conflicts and updated

premik91 commented:
So I should remove userlist.html and move the code to home.html?

premik91 commented:
Should I make panel reusable like this? 

.userPanel {
    width: 260px;
}

.panelHeader {
    color: #FCFCFC;
    height: 20px;
    text-align: center;
    background-color: #68AE2C;
    cursor: pointer;
}

---

< div class="userPanel" >
    < div class="panelHeader" title="{% trans "Show/Hide Users" %}">{% trans "Active Users" %}< / div>

mitar commented:
No, it is OK that it is in `userlist.html`. Just include `userlist.html` in `home.html`.

mitar commented:
For now leave panels as they are. We will generalize them later. (There is ticket for that.)

premik91 commented:
It already is included in home.html

mitar commented:
Aha. OK.

Commit:
Removed checkbox from html

mitar commented:
And the rest? JavaScript?

mitar commented on piplmesh/frontend/templates/userlist.html:
```diff
@@ -1,7 +1,14 @@
         {% endfor %}
     </ul>
+    <form action="">
+        <img src="{{ STATIC_URL }}piplmesh/images/magnifying_glass.png" alt="magnifying_glass" />
```
Alt should be some translatable string ("Search users" maybe?).

premik91 replied:
Fixed

mitar commented on piplmesh/frontend/tasks.py:
```diff
@@ -27,7 +27,12 @@ def check_online_users():
+                    'user': {
+                        'username': user.username,
+                        'profile_url': urlresolvers.reverse('user', kwargs={'username': user.username}),
+                        # TODO get user image
```
Format this TODO into format of other:

```
# TODO: Get real user image
```

premik91 replied:
Fixed

mitar commented on piplmesh/frontend/tasks.py:
```diff
@@ -27,7 +27,12 @@ def check_online_users():
+                        'username': user.username,
+                        'profile_url': urlresolvers.reverse('user', kwargs={'username': user.username}),
+                        # TODO get user image
+                        'image_url': '',
```
For now you can use our default user image we already have. So that there will be something. Just hard-code it (but still build image path dynamically).

premik91 replied:
You mean like I had it before? Like 

'image_url': '../images/logo.png'

mitar replied:
No, you use `django.contrib.staticfiles.storage.staticfiles_storage.url(path)`.

premik91 replied:
I was just searching for something like that, thx

igypigi commented:
i'm working on JavaScript right now.

mitar commented on piplmesh/frontend/static/piplmesh/css/style.css:
```diff
@@ -400,3 +399,52 @@ input.gsc-search-button {
+    float: left;
+}
+
+#userlist .userInfo {
```
_userInfo_ -> _info_?

mitar commented on piplmesh/frontend/static/piplmesh/css/style.css:
```diff
@@ -400,3 +399,52 @@ input.gsc-search-button {
+
+#userlist img {
+    height: 25px;
+    width: 25px;
```
I think ticket is saying that user images should be 20x20 px? Is that too small? I don't think so.

premik91 replied:
I thought they were, that's why I made them 25x25. I thought they were much better and clearer this size. I thought they looked good and just the correct size? http://shrani.najdi.si/?3x/Gw/3KRkNgmi/userlist.png

mitar replied:
Hm. In fact Facebook has 28x28 in chat. And the one in header menu is 23x23. I do not know where I found 20x20. But they have only 50x50 resized. The rest they resize with CSS. So what should we have?

premik91 replied:
You tell me :). At the begining I had them 20x20, but they seemed to small and the addtitional user info didn't look so good next to the image. That was also one of the reasons I made them bigger.

mitar replied:
OK, so you are saying 25x25 is OK, then use it.

premik91 replied:
Yes I think it is okay, but that can easily be changed if necessary

mitar commented:
Good. You understand what I want from you in JavaScript, yes?

igypigi commented:
i think so. but i dont know how to move django/python code for initial display of users already online to javascript?

mitar commented:
The point is that you are currently redrawing the list of users on every update. And that you begin with an empty list (or dict) of users in JavaScript. But why wouldn't you start with a filled one, filled with currently known online users. And then just redraw on document ready. And then keep it updated on pushed updates.

So how can you get list of currently known online users into a JavaScript variable:
- you create a custom Django view, which dynamically create JavaScript code which defines this variable and load it through HTML script tag
- you output such code in home template itself, similar to how we are doing [it here](https://github.com/wlanslovenija/nodewatcher/blob/master/nodewatcher/frontend/templates/nodes/map.html#L90)

Heh, we are even doing this [already in PiplMesh](https://github.com/wlanslovenija/PiplMesh/blob/master/piplmesh/frontend/templates/base.html#L14).

mitar commented:
Once you have data in JavaScript variable - you know how to draw this into HTML, no? :-)

igypigi commented:
i will try that. thanks :)

Commit:
JS fixes by igypigi

Commit:
Fixed TODO and css naming

Commit:
Fixed image url and alt attribute in userlist.html

premik91 commented:
As I have read javascript isn't supported yet by Django(but that could be old documentation)? So the way igypigi did is wrong?

premik91 commented:
Nope doesn't work. I will inform him.

premik91 commented:
I've also noticed that when you log in, the text "HOME PAGE CONTENT HERE " changes style. But this should be in another Issue.

mitar commented:
> As I have read javascript isn't supported yet by Django

What you mean by that? Supported for what?

premik91 commented:
I meant using django code in javascript?

mitar commented:
What means "django code in JavaScript"?

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,68 @@
-		// TODO: Improve escape and use more suitable :contains method
-		$('#userlist li:contains(' + escape(data.username) + ')').remove();
-	}
+var onlineUsers = {
```
Yes, this is wrong. This should go to Django templates. JavaScript files are static files and are not processed by Django. They could be. But we are not doing that with this file.

You can:
- move this to `home.html` template
- create a view returning this JavaScript code
- you can do the later through templates, if you want

mitar replied:
And this?

premik91 replied:
Much done, but still some fixes necessary. Should we move all functions inside the document.ready function?
<code>
$(document).ready(function () {
    $.updates.registerProcessor('home_channel', 'userlist', updateUserList);
    /\* ------------------ USER PANEL ---------------------------- */
    var searchUsers = '';

```
function User(data) {
    var self = this;
    $.extend(self, data.user);
}
```

</code>

mitar replied:
No, have objects outside and then methods inside those objects. For now leave as it is. We will restructure code when we will have more of it.

premik91 commented:
I was talking about this.
var onlineUsers = {
    {% for user in online_users %}
        {{ user.username }}: {{ user }} {% if not forloop.last %}, {% endif %}
    {% endfor %}
};
It is as simple as that, does this work or not. This is what I said was "Django code".

mitar commented:
Yes, I commented on this in the code.

premik91 commented:
Yes I saw, we were probably commenting at the same time :)

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,68 @@
+}
+
+function redrawUserList() {
+    var keys = [];
```
`usernames` maybe?

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,68 @@
+        if (searchUsers === "" || key.indexOf(searchUsers) != -1) {
+            user = onlineUsers[key];
+            var li = $('<li/>');
+            var image = $('<img/>').attrs({'src':user.image_url, 'alt':'User image'});
```
Translatable string. [Check this](https://docs.djangoproject.com/en/dev/topics/i18n/translation/#internationalization-in-javascript-code).

mitar replied:
Space after `:`.

mitar replied:
You can also wrap dict in a similar way we do it in Python.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,68 @@
+            var image = $('<img/>').attrs({'src':user.image_url, 'alt':'User image'});
+            li.append(image);
+            li.append(user.username);
+            var div = $('<div/>').attrs({'class':'info'});
```
Space.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,68 @@
+function updateUserList(data) {
+    var user = new User(data.user);
+    if (data.action === 'JOIN') {
+        if (!onlineUsers.hasOwnProperty(user.username)) {
```
No, don't check. Maybe the new update has some new information (new user image, for example)?

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,68 @@
+        }
+    } else if (data.action === 'PART') {
+        var index = onlineUsers.indexOf(data);
+        if (onlineUsers.hasOwnProperty(user.username)) {
```
`onlineUsers[user.username]` is good enough.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,68 @@
+    $.updates.registerProcessor('home_channel', 'userlist', updateUserList);
+
+    $(".panel .header").click(function () {
+        $(this).next("ul").slideToggle("slow");
```
Do it fast.

Please use `'` everywhere for constant strings.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,68 @@
+        $(this).next("ul").slideToggle("slow");
+    });
+
+    $("#search_users").keyup(function () {
```
`keyup`? Not `change`? What if I paste something in? Doe sit still work?

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,68 @@
+    });
+
+    $("#search_users").keyup(function () {
+        searchUsers = $(this).val();
```
I don't see the reason to use one more global variable. Why we cannot simply `$('#search_users').val()` in the code above when we need its content?

mitar commented on piplmesh/frontend/tasks.py:
```diff
@@ -27,7 +27,12 @@ def check_online_users():
+                        'profile_url': urlresolvers.reverse('user', kwargs={'username': user.username}),
+                        # TODO: Get real user image
+                        'image_url': django.contrib.staticfiles.storage.staticfiles_storage.url('../images/logo.png'),
+                    },
```
Of course you should import this in a way to have a shorted function call!

mitar replied:
And no, it is not correct to use `../images/logo.png`. Do we use such path in Django templates for static files?

premik91 replied:
Is path now okay? Or should it be just piplmesh/images/logo.png

mitar commented on piplmesh/frontend/templates/userlist.html:
```diff
@@ -1,7 +1,10 @@
+{% load i18n %}
+
+<div class="panel">
+    <div class="header" title="{% trans "Show/Hide Users" %}">{% trans "Active Users" %}</div>
```
_Active_ -> _Online_

Commit:
Fixed import and url, Active -> Online

mitar commented:
Waiting for JavaScript ...

mitar commented:
BTW, fixes from igypigi should be committed by him. Otherwise he will not get any credit for his work towards his course mark.

mitar commented on piplmesh/frontend/tasks.py:
```diff
@@ -27,7 +29,12 @@ def check_online_users():
+                        'username': user.username,
+                        'profile_url': urlresolvers.reverse('user', kwargs={'username': user.username}),
+                        # TODO: Get real user image
+                        'image_url': django.contrib.staticfiles.storage.staticfiles_storage.url('/piplmesh/images/logo.png'),
```
Again, why full import of the function? And please, just try what is the output of your call and see if it is right.

premik91 replied:
Now it all works and gives the expected output.

Commit:
Fixed import and path to static files

mitar commented on piplmesh/frontend/tasks.py:
```diff
@@ -27,7 +29,12 @@ def check_online_users():
+                        'username': user.username,
+                        'profile_url': urlresolvers.reverse('user', kwargs={'username': user.username}),
+                        # TODO: Get real user image
+                        'image_url': staticfiles_storage.url('/static/piplmesh/images/logo.png'),
```
No, it is still not right. You should not hard-code final URL in. Then you have not done anything. :-)

premik91 replied:
haha maybe it would be faster to do TODO :D. 
So I should just leave 'piplmesh/images/logo.png' and add {{ STATIC_URL }} in js?

mitar replied:
Nooo. First, template tags do not work in JS. Because of that we will construct URL in Python.

In fact, concatenating STATIC_URL like that is not the best way to do it. We should use `static` template tag. Please update also templates.

[Read here more](https://docs.djangoproject.com/en/dev/howto/static-files/#staticfiles-in-templates) on the subject.

And then when you will know how to do that in templates, you can call the same function which is used in `static` template tag to get also URL in the code.

premik91 replied:
Is this okay now? Static tag works. Should I do this for all templates?

mitar replied:
Yes.

premik91 replied:
I am having problems here, because if I change this
<code>"{{ STATIC_URL }}piplmesh/images/logo-{{ LANGUAGE_CODE }}.png"</code>
into this
<code>"{% static "piplmesh/images/logo-{{ LANGUAGE_CODE }}.png" %}"</code>,
the _{{ LANGUAGE_CODE }}_ variable does't work. Should I use js for this(I think that might work)?

mitar replied:
Of course not. Why use JS?

Anyway, you will have to add new context variable in our context processor and construct the filepath in Python. And use that then in template. Read second paragraph:

https://docs.djangoproject.com/en/dev/howto/static-files/#with-a-template-tag

premik91 replied:
It works but you will certainly hate the way I added strings. :)

Commit:
Fixed image_url

premik91 commented:
Is this what you meant? I read all about static files, but about staticfiles_storage there are very little examples. I also googled it and since this works I don't know what else to do. I can't say what igypigi is waiting for...

Commit:
Merge remote-tracking branch 'upstream/master' into Issue-#141--team5

mitar commented:
Yes. Also do the same in templates. So instead of using that variable, use tag. (Import it from `staticfiles` library.)

Commit:
jq minor fixes

premik91 commented:
I don't understand what to do. In templates tag is already used.
<code>< img src="{{ STATIC_URL }}piplmesh/images/magnifying_glass.png" /></code>

mitar commented:
Load `staticfiles` library and use `static` template tag. That is not template tag, but variable. We are concatenating strings manually, it is better through tag.

igypigi committed:
fixes from igypigi and premik91

fixed js,
added script for initial display of users,
changed user model, updated tasks.py

Commit:
Removed unused import

Commit:
Fixed import

Commit:
Minor js fixes

Commit:
Load Tag for static files

mitar commented on piplmesh/account/models.py:
```diff
@@ -46,3 +48,7 @@ class User(auth.User):
+
+    profile_url = urlresolvers.reverse('user', kwargs={'username': user.username})
+    # TODO: Get real user image
+    image_url = staticfiles_storage.url('piplmesh/images/logo.png')
```
Change this into two methods on User class, `get_profile_url` and `get_image_url`.

mitar commented on piplmesh/account/models.py:
```diff
@@ -46,3 +48,7 @@ class User(auth.User):
     connection_last_unsubscribe = mongoengine.DateTimeField()
     is_online = mongoengine.BooleanField(default=False)
+
+    profile_url = urlresolvers.reverse('user', kwargs={'username': user.username})
```
`user` variable is defined here??? Move to method and use `self`.

premik91 replied:
Ups. Fixed :)

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,61 @@
+
+function User(data) {
+    var self = this;
+    $.extend(self, data.user);
```
Ups. Use `data` here. And pass into constructor `data.user`. It is much better (now that we have a separate dict only for user fields).

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,61 @@
+}
+
+function updateUserList(data) {
+    var user = new User(data.user);
```
Oh, you are already passing `data.user` here. :-)

mitar commented on piplmesh/frontend/templates/userlist.html:
```diff
@@ -1,7 +1,28 @@
+        </script>
+{% endblock %}
+
+{% load staticfiles %}
```
Move all loading to one template tag (you can list multiple libraries to load) on the top.

mitar commented on piplmesh/frontend/templates/userlist.html:
```diff
@@ -1,7 +1,28 @@
-    </ul>
+{% load i18n %}
+
+{% block js %}
```
Good. But also first add parent content `{{ block.super }}`, because maybe some other template has already used this block, to not lose its JavaScript code.

mitar replied:
See documentation on `block` tag for more information. Or documentation on template inheritance.

premik91 replied:
Fixed

premik91 replied:
I've read the documentation and tried block.super, but it doesn't work. <code>Error: 'Caught AttributeError while rendering: 'BlockNode' object has no attribute 'context' '</code>
Than I googled the problem and found this and I think it makes sense: 
http://stackoverflow.com/questions/6566347/django-templates-using-block-super-in-included-template-fails-exception

mitar replied:
Then move the whole this `js` block to `home.html` template.

mitar commented on piplmesh/frontend/templates/userlist.html:
```diff
@@ -1,7 +1,28 @@
+                        'profile_url': '{{ user.profile_url }}',
+                        'image_url': '{{ user.image_url }}'
+                    }
+                    {% if not forloop.last %}, {% endif %}
```
Unnecessary space after comma.

premik91 replied:
Fixed

mitar commented on piplmesh/frontend/templates/userlist.html:
```diff
@@ -1,7 +1,28 @@
+{% load i18n %}
+
+{% block js %}
+        <script type="text/javascript">
```
Indentation (too much).

premik91 replied:
Fixed

mitar commented on piplmesh/frontend/templates/userlist.html:
```diff
@@ -1,7 +1,28 @@
+            var onlineUsers = [
+                {% for user in online_users %}
+                    {
+                        'username': '{{ user.username }}',
```
You should always escape values. Django template variables are escaped automatically only for HTML. Not for JavaScript. There is a template filter for this.

premik91 replied:
<code>{{ user.username|escapejs }}</code>
Like this?

mitar replied:
Yes!

mitar commented on piplmesh/frontend/tasks.py:
```diff
@@ -7,6 +7,7 @@
 from pushserver.utils import updates
 
 from piplmesh.account import models
+
```
No.

premik91 replied:
Fixed

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,61 @@
+        if (searchUsers === '' || key.indexOf(searchUsers) != -1) {
+            user = onlineUsers[key];
+            var li = $('<li/>');
+            var image = $('<img/>').attrs({'src': user.image_url, 'alt': 'User image'});
```
Translatable string. Check how translation is done in JavaScript.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,61 @@
+        onlineUsers[user.username] = user;
+        redrawUserList();
+    } else if (data.action === 'PART') {
+        var index = onlineUsers.indexOf(data);
```
No need for this?

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,61 @@
+    });
+
+    $('#search_users').change(function () {
+        searchUsers = $(this).val();
```
Why do we have this variable? Why not simply call `$('#search_users').val()` in `redrawUserList`?

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,61 @@
+    });
+    $('#userlist').empty();
+    for (key in keys) {
+        if (searchUsers === '' || key.indexOf(searchUsers) != -1) {
```
This should be probably case-insensitive search. Fix or add TODO.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,61 @@
+
+function redrawUserList() {
+    var keys = [];
+    for (key in onlineUsers) {
```
OK, use [$.each](http://api.jquery.com/jQuery.each/) here.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,61 @@
+        return 0;
+    });
+    $('#userlist').empty();
+    for (key in keys) {
```
OK, use [$.each](http://api.jquery.com/jQuery.each/) here.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,61 @@
+function redrawUserList() {
+    var keys = [];
+    for (key in onlineUsers) {
+        if (onlineUsers.hasOwnProperty(key)) {
```
No need for this then.

Commit:
First fixes

Commit:
Added block.super

Commit:
Removed block.super

Commit:
load staticfiles in templates

Commit:
fixed imports

mitar commented on piplmesh/account/models.py:
```diff
@@ -7,6 +7,10 @@
 from piplmesh.account import fields
 
+from django.contrib.staticfiles.storage import staticfiles_storage
+
```
No empty line between imports from same package.

mitar commented on piplmesh/frontend/templates/userlist.html:
```diff
@@ -1,7 +1,27 @@
+                'image_url': '{{ user.get_image_url|escapejs }}'
+            }
+            {% if not forloop.last %},{% endif %}
         {% endfor %}
```
Move this line up, to the end of previous line.

Commit:
Fixes

Removed empty line, added new context variable in our context
processor, removed variables in home.js, base.html load static files,
moved js block to home.html

Commit:
Line fix

Commit:
Fixes

Commit:
Added empty lines after load

igypigi committed:
fixes

fixed User constructor,
iterate over array and dictionary with $.each()

igypigi committed:
fixed

fixed user search to be case insensitive, keys of onlineUsers are now
UpperCase

Commit:
Final fix

Commit:
Compilemessages run

Commit:
Updated documentation

premik91 commented:
Waiting for approval :)

Commit:
Fixed documentation

mitar commented on docs/translation.rst:
```diff
@@ -19,14 +20,16 @@ sure your Django applications you want translated have a directory called
 
 When all string are marked for translation and settings are prepared run
-command::
+command(the second one if you need JavaScript files translated)::
```
No need for this comment, because we want to translate everything, of course.

premik91 replied:
Fixed

mitar commented on docs/translation.rst:
```diff
@@ -19,14 +20,16 @@ sure your Django applications you want translated have a directory called
+command(the second one if you need JavaScript files translated)::
 
     ../manage.py makemessages -l <language code>
+    ../manage.py makemessages -l <language code> -d djangojs
```
Change to:

```
../manage.py makemessages -l <language code> -d djangojs
```

premik91 replied:
Fixed

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,57 @@
+
+function redrawUserList() {
+    var keys = [];
+    $.each(onlineUsers, function(key, value){
```
For anonymous function put spaces around `()`.

premik91 replied:
Fixed

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,57 @@
+    $.each(onlineUsers, function(key, value){
+        keys.push(key);
+    });
+    keys.sort(function (key1, key2) {
```
Same as here.

mitar commented on piplmesh/frontend/context_processors.py:
```diff
@@ -9,10 +9,13 @@ def global_vars(request):
         'SEARCH_ENGINE_UNIQUE_ID': settings.SEARCH_ENGINE_UNIQUE_ID,
 
+        # Variables
+        'logo_url': "piplmesh/images/logo-%s.png" % (settings.LANGUAGE_CODE,),
```
Nop. Don't read from `settings` (that is just default), you have to read from context.

premik91 replied:
I don't understand what to do. You said I should put it like this :)

mitar replied:
It is very good what you did, only the value you put into the string is wrong. `LANGUAGE_CODE` in the template context is not the same as `settings.LANGUAGE_CODE`, but depends on the current user language. `settings.LANGUAGE_CODE` is only fallback value for that value in template context.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,57 @@
+function updateUserList(data) {
+    var user = new User(data.user);
+    if (data.action === 'JOIN') {
+        onlineUsers[user.username.toUpperCase()] = user;
```
Good idea. But go to lower case. ;-)

premik91 replied:
Fixed

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,57 @@
+        onlineUsers[user.username.toUpperCase()] = user;
+        redrawUserList();
+    } else if (data.action === 'PART') {
+        if (onlineUsers[user.username]) {
```
But then you have to convert to same case everywhere! And this means it is easy to break.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,57 @@
-	}
+function User(data) {
+    var self = this;
+    $.extend(self, data);
```
I suggest that you add:

```
self._key = self.username.toLowerCase();
```

And then use `user._key` for keys.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,57 @@
+        return 0;
+    });
+    $('#userlist').empty();
+    var searchUsers = $('#search_users').val().toUpperCase();
```
Replace this and second line. This goes more with the next block of code.

mitar commented on piplmesh/urls.py:
```diff
@@ -10,6 +10,11 @@
 v1_api.register(resources.PostResource())
 
+js_info_dict = {
+    'domain': 'djangojs',
```
This is already default.

premik91 replied:
Fixed

mitar commented on piplmesh/urls.py:
```diff
@@ -10,6 +10,11 @@
 
+js_info_dict = {
+    'domain': 'djangojs',
+    'packages': ('piplmesh.frontend',),
```
Add also `django.conf`.

premik91 replied:
Fixed

mitar commented on piplmesh/urls.py:
```diff
@@ -10,6 +10,11 @@
+js_info_dict = {
+    'domain': 'djangojs',
+    'packages': ('piplmesh.frontend',),
+    }
```
Indent.

premik91 replied:
Fixed

mitar commented on piplmesh/urls.py:
```diff
@@ -37,4 +42,6 @@
     # RESTful API
     url(r'^api/', include(v1_api.urls)),
+
+    (r'^jsi18n/$', 'django.views.i18n.javascript_catalog', js_info_dict),
```
And where do you use it?

premik91 replied:
Fixed?

mitar commented on piplmesh/frontend/templates/base.html:
```diff
@@ -1,14 +1,14 @@
+        <link type="image/x-icon" href="{% static "piplmesh/images/favicon.ico" %}" rel="icon" />
+        <link rel="stylesheet" href="{% static "piplmesh/css/style.css" %}" type="text/css" media="screen" />
+        <script src="{% static "piplmesh/jquery/jquery.js" %}" type="text/javascript"></script>
+        <script type="text/javascript" src="{% static "pushserver/updates.js" %}"></script>
```
Move `type` to the end of the tag.

premik91 replied:
Fixed

Commit:
Fixes

Commit:
Moved type to the end of tag

mitar commented on docs/translation.rst:
```diff
@@ -8,6 +8,7 @@ To make your strings translatable edit them as follows:
 
 - in Django template: ``{% trans "Some string" %}``
```
_template_ -> _templates_

mitar commented on piplmesh/frontend/templates/home.html:
```diff
@@ -2,6 +2,23 @@
+<script type="text/javascript">
+    /* <![CDATA[ */
+    var onlineUsers = [
+    {% for user in online_users %}
```
Indentation. Every block indent.

mitar commented on piplmesh/urls.py:
```diff
@@ -10,11 +10,16 @@
 v1_api.register(resources.PostResource())
 
+js_info_dict = {
+        'packages': ('piplmesh.frontend', 'django.conf',),
```
Order by alphabet. Split into lines.

mitar commented on piplmesh/urls.py:
```diff
@@ -10,11 +10,16 @@
 
+js_info_dict = {
+        'packages': ('piplmesh.frontend', 'django.conf',),
+    }
```
Indentation.

mitar commented on piplmesh/locale/sl/LC_MESSAGES/djangojs.po:
```diff
@@ -0,0 +1,22 @@
```
Compare with `django.po` and sync style/header.

mitar commented on piplmesh/frontend/templates/base.html:
```diff
@@ -1,14 +1,14 @@
-        <script src="{{ STATIC_URL }}piplmesh/jquery/jquery.js" type="text/javascript"></script>
-        <script type="text/javascript" src="{{ STATIC_URL }}pushserver/updates.js"></script>
-        <script src="{{ STATIC_URL }}piplmesh/js/home.js" type="text/javascript"></script>
+        <link type="image/x-icon" href="{% static "piplmesh/images/favicon.ico" %}" rel="shortcut icon" />
```
Groan. Should I really point everything out. You do not notice that also in just few tags above also things are not ordered in the same way.

premik91 replied:
Well it isn't exactly the same. And it could be just the other way around, I would've change it and you would groan why did I do something you didn't want(it is not that I didn't notice it).

mitar replied:
It is not consistent - it is the same. I would not groan if you would make things more consistent.

Commit:
Fixes

Commit:
Getting the right value from context

Commit:
Updated fixed conflicts

igypigi committed:
fixes

premik91 commented:
This is now it.

mitar commented on docs/translation.rst:
```diff
@@ -21,12 +22,12 @@ sure your Django applications you want translated have a directory called
 When all string are marked for translation and settings are prepared run
 command::
 
-    ../manage.py makemessages -l <language code>
```
No, you have to run both, no? Without `-d` (for Python and template) and with `-d` (for JavaScript).

mitar replied:
TODO.

premik91 replied:
Yes you have to run both, but in previous comment you said to delete the other one(I might have understood you wrong).

mitar replied:
Yes, you need both. So have both here.

(I don't find such comment. But it does not matter.)

mitar replied:
TODO.

premik91 replied:
http://shrani.si/f/3p/zV/4F3HF7ai/f.png
I thought you meant to replace both with one.

mitar replied:
No, you had some `-a` parameter there.

mitar commented on piplmesh/frontend/context_processors.py:
```diff
@@ -9,10 +9,14 @@ def global_vars(request):
         'SEARCH_ENGINE_UNIQUE_ID': settings.SEARCH_ENGINE_UNIQUE_ID,
 
+        # Variables
+        'logo_url': "piplmesh/images/logo-%s.png" % (request.LANGUAGE_CODE),
```
It is better to use `translation.get_language()` to get language code. It is used also in `django.core.context_processors.i18n`, where template's `LANGUAGE_CODE` is set. So we should probably use the same. `request.LANGUAGE_CODE` is also set from `translation.get_language()`, so the value is the same, but it assumes that we always be using `django.middleware.locale.LocaleMiddleware` which sets it. So it is better to have one less assumption/dependency.

Also, put comma before last `)`, because you want to force tuple.

mitar replied:
TODO.

premik91 replied:
Yes the main problem was, I didn't even suspect that igypigi will do the fixes, so I didn't sync the branch, and didn't have igypigi's fixes. And than Github crashed and so on. I hope there are no interferences.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,58 @@
+    var searchUsers = $('#search_users').val().toLowerCase();
+    $.each(keys, function(index, key){
+        if (searchUsers === '' || key.indexOf(searchUsers) !== -1) {
+            user = onlineUsers[key];
```
`var user =`, not just `user =`.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,58 @@
+        redrawUserList();
+    } else if (data.action === 'PART') {
+        if (onlineUsers[user._key]) {
+            delete onlineUsers[user.username.toLowerCase()];
```
Use `user._key`.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,58 @@
+    $('#userlist').empty();
+
+    var searchUsers = $('#search_users').val().toLowerCase();
+    $.each(keys, function(index, key){
```
Spaces around `(index, key)`. See how we have other anonymous functions.

mitar commented on piplmesh/frontend/templates/base.html:
```diff
@@ -1,14 +1,14 @@
-        <script src="{{ STATIC_URL }}piplmesh/js/home.js" type="text/javascript"></script>
+        <link href="{% static "piplmesh/images/favicon.ico" %}" rel="shortcut icon" type="image/x-icon" />
+        <link href="{% static "piplmesh/images/favicon.ico" %}" rel="icon" type="image/x-icon" />
+        <link rel="stylesheet" href="{% static "piplmesh/css/style.css" %}" type="text/css" media="screen" />
```
Put `href` before `rel`?

mitar commented on piplmesh/frontend/templates/home.html:
```diff
@@ -2,6 +2,23 @@
+    /* <![CDATA[ */
+    var onlineUsers = [
+        {% for user in online_users %}
+        {
```
Indent once. Do not care about output look. Care about template readability.

mitar commented on piplmesh/frontend/templates/home.html:
```diff
@@ -2,6 +2,23 @@
 {% block fulltitle %}wlan slovenija{% endblock %}
 
+{% block js %}
+{{ block.super }}
```
Indent inside the block.

mitar commented on piplmesh/urls.py:
```diff
@@ -10,11 +10,19 @@
 
     url(r'^search/', frontend_views.SearchView.as_view(), name='search'),
     url(r'^i18n/', include('django.conf.urls.i18n')),
+    (r'^jsi18n/$', 'django.views.i18n.javascript_catalog', js_info_dict),
```
Put `url` at the beginning, to be consistent.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -1,13 +1,58 @@
+    if (data.action === 'JOIN') {
+        onlineUsers[user._key] = user;
+        redrawUserList();
+    } else if (data.action === 'PART') {
```
`else if` into new line.

mitar commented on piplmesh/frontend/templates/form/form.html:
```diff
@@ -9,31 +9,31 @@
```
What did you do with this file? Broke indentation? I see that you cleaned some things. But also broke some things. So we indent always. For template or HTML block/tag.

mitar replied:
Please improve.

premik91 replied:
I did nothing here? It is the same as it is now?

mitar replied:
Now is OK.

mitar replied:
Aaaagh. Again it shows a lot of changes in this file. I am not sure what are you doing! ;-)

premik91 replied:
Haha, me neither I never even had it open. Will fix it back. The problem was I didn't sync igypigi's fixes, and than there was somekind of bug in Github for Mac and it crashed in the middle of commit , ...

premik91 replied:
It didn't even show in diff

premik91 replied:
But I see no problems still? :D I just copied the current version in, and nothing changed

mitar replied:
I don't know what to tell you. You do see it in pull request diff?

premik91 replied:
Look you know it is not hard to make another commit to fix it. But I just copied the current version in my file, and github sees no changes. Why would I lie to you? I will do somethin else then, make a stupid commit and then another commit with current code.

mitar replied:
I am not saying that you are lying. I am saying that also I don't know what is happening. ;-)

What does `git log` says about the file? When it was last time changed? Can you checkout that version of the file? Where have you copied current version from? How do you copy it? Maybe close your IDE. ;-)

premik91 replied:
I think now is okay. I deleted the file and made a new one. Github than recognized as a new file(?!) and was okay.

mitar commented:
It looks very good now.

igypigi committed:
fixes

igypigi committed:
fixes

mitar commented on piplmesh/frontend/context_processors.py:
```diff
@@ -9,10 +9,14 @@ def global_vars(request):
 
+        # Variables
+        'logo_url': "piplmesh/images/logo-%s.png" % (request.LANGUAGE_CODE),
+        'redirect_to': request.REQUEST.get(auth.REDIRECT_FIELD_NAME),
```
We have renamed this field (`request_get_next`). Please do not reintroduce it.

mitar replied:
TODO.

mitar commented on piplmesh/frontend/templates/home.html:
```diff
@@ -2,6 +2,24 @@
+                    'profile_url': '{{ user.get_profile_url|escapejs }}',
+                    'image_url': '{{ user.get_image_url|escapejs }}'
+                }
+                {% if not forloop.last %},{% endif %}
```
You like this on its own line? You moved it again there. Then OK, let it be. ;-)

igypigi replied:
I think it is easier to understand the code and it looks better this way. :)

mitar replied:
OK.

Commit:
Fixes

Commit:
FIxes

Commit:
Merge remote-tracking branch 'upstream/master' into Issue-#141--team5

mitar commented on docs/translation.rst:
```diff
@@ -22,11 +23,12 @@ When all string are marked for translation and settings are prepared run
 *(You should replace <language code> with your preferred language code).*
 Example for Slovenian language::
 
-    ../manage.py makemessages -l sl
```
Don't remove this line.

premik91 replied:
omg, I didn't. Arrrrgh. I was trying to revert my commit and it crahed again.

mitar commented on piplmesh/frontend/context_processors.py:
```diff
@@ -9,10 +10,14 @@ def global_vars(request):
 
+        # Variables
+        'logo_url': "piplmesh/images/logo-%s.png" % (translation.get_language(),),
+        'redirect_to': request.REQUEST.get(auth.REDIRECT_FIELD_NAME),
```
This variable is not needed anymore! Why are you putting it back?

Commit:
Fix

Commit:
unfix

Commit:
Fix

Commit:
Fresh copy, deleted old

mitar commented:
Done. Congrats!
